security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java

```diff
@@ -212,52 +211,85 @@ private void addRequestAsHeader(ServerRequest req, ServerResponse res) {
-        formValues.putSingle("code", code);
-        formValues.putSingle("redirect_uri", oidcConfig.redirectUriWithHost());
-
-        Response response = oidcConfig.tokenEndpoint().request()
-                .accept(MediaType.APPLICATION_JSON_TYPE)
-                .post(Entity.form(formValues));
-
-        if (response.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL) {
-            JsonObject jsonResponse = response.readEntity(JsonObject.class);
-            String tokenValue = jsonResponse.getString("access_token");
-            //redirect to "state"
-            String state = req.queryParams().first(STATE_PARAM_NAME).orElse(DEFAULT_REDIRECT);
-            res.status(Http.Status.TEMPORARY_REDIRECT_307);
-            if (oidcConfig.useParam()) {
-                if (state.contains("?")) {
-                    state = state + "&" + oidcConfig.paramName() + "=" + tokenValue;
-                } else {
-                    state = state + "?" + oidcConfig.paramName() + "=" + tokenValue;
-                }
-            }
+        WebClient webClient = oidcConfig.appWebClient();
+
+        FormParams.Builder form = FormParams.builder()
+                .add("grant_type", "authorization_code")
+                .add("code", code)
+                .add("redirect_uri", oidcConfig.redirectUriWithHost());
+
+        WebClientRequestBuilder post = webClient.post()
+                .uri(oidcConfig.tokenEndpointUri())
+                .accept(io.helidon.common.http.MediaType.APPLICATION_JSON);
+
+        oidcConfig.updateRequest(OidcConfig.RequestType.CODE_TO_TOKEN,
+                                 post,
+                                 form);
 
-            state = increaseRedirectCounter(state);
-            res.headers().add(Http.Header.LOCATION, state);
+        OidcConfig.postJsonResponse(post,
+                                    form.build(),
+                                     json -> processJsonResponse(req, res, json),
+                                    (status, errorEntity) -> processError(res, status, errorEntity),
+                                    (t, message) -> processError(res, t, message))
+                .ignoreElement();
 
-            if (oidcConfig.useCookie()) {
-                res.headers()
-                        .add("Set-Cookie", oidcConfig.cookieName() + "=" + tokenValue + oidcConfig.cookieOptions());
+    }
+
+    private String processJsonResponse(ServerRequest req, ServerResponse res, JsonObject json) {
+        String tokenValue = json.getString("access_token");
+        //redirect to "state"
+        String state = req.queryParams().first(STATE_PARAM_NAME).orElse(DEFAULT_REDIRECT);
+        res.status(Http.Status.TEMPORARY_REDIRECT_307);
+        if (oidcConfig.useParam()) {
+            if (state.contains("?")) {
+                state = state + "&" + oidcConfig.paramName() + "=" + tokenValue;
+            } else {
+                state = state + "?" + oidcConfig.paramName() + "=" + tokenValue;
```

Comment:
Could be little shorter
`state = state + (state.contains("?") ? "&" : "?") + oidcConfig.paramName() + "=" + tokenValue;`